### PR TITLE
Add declaration-empty-line-before

### LINKED ISSUE #1573: declaration-empty-line-before
- Name: `declaration-empty-line-before`
- Primary option: `"always"|"never"`
- Secondary options: `{ "except": ["first-nested", "after-comment", "after-declaration"] }`
- Messages: 
  - Rejected: "Unexpected empty line before declaration"
  - Expected: "Expected empty line before declaration"
- Section: "Declaration"

Given:

```js
[ "always", { "except": ["first-nested", "after-comment", "after-declaration"] }]
```

The following is valid:

```css
a {
  prop: value; /* "always" except first-nested */
  prop2: value; /* "always" except after-declaration */

  /* comment */
  prop3: value; /* "always" except after-comment */

  @extends "x";

  prop4: value; /* primary option is "always" */

   & b {
     prop: value;
   }

   prop5: value;  /* primary option is "always" */
}
```

Given:

```js
"never"
```

The following is valid:

```css
a {
  prop: value; /* primary option is "never" */
  prop2: value; /* primary option is "never" */

  /* comment */
  prop3: value; /* primary option is "never" */

  @extends "x";
  prop4: value; /* primary option is "never" */

   & b {
     prop: value;
   }
   prop5: value; /* primary option is "never" */
}
```

Ref: https://github.com/stylelint/stylelint/issues/1500#issuecomment-229300139

### PR DESCRIPTION
**Pull request by jeddy3**
Closes #1573

Doing this uncovered a few issues with the approach in `custom-property-empty-line-before`:

1. Using `walkDecls` and `walkRules` together meant declarations within nested rules were being checked twice.
2. We hadn't considered single-line blocks...

I've gone with just using `walkDecls` to address the former.

And I've gone with an `ignore: ["inside-single-line-block"]` secondary option (ok name?) to address the latter. 

This is at odds with how I've dealt with single-line blocks within `block-closing-brace-empty-line-before`. There I've used an `"always-multi-line"`, rather than `"always`" option. As this is inline with the other `block-*-brace` rules.

I had considered scoping `declaration-empty-line-before` to the declaration block e.g. `declaration-block-declaration-empty-line-before`. This would allow us to use `"always-multi-line"` in a consistent way. The naming would also be consistent with things like `declaration-block-single-line-max-declarations`. *However*, this would be at odds with `rule-nested-empty-line-before` and `at-rule-empty-line-before`.

Long story short, I think adding an `ignore: ["inside-single-line-block"]` option to this and `custom-property-empty-line-before` is the best compromise on consistency.

If we're happy with this, I'll retrofit these changes to `custom-property-empty-line-before`.

### CODE REVIEW

File: src/rules/declaration-empty-line-before/index.js [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+      },
+      optional: true,
+    })
+    if (!validOptions) { return }
+
+    root.walkDecls(decl => {
+      const { prop, parent } = decl
+
+      if (!isStandardSyntaxDeclaration(decl)) { return }
+      if (!isStandardSyntaxProperty(prop)) { return }
```
Comment (davidtheclark) [inline src/rules/declaration-empty-line-before/index.js]: Are these two checks necessary? What irregularities do they prevent?

Comment (davidtheclark) [inline src/rules/declaration-empty-line-before/index.js]: I guess I have a concern that people will want not distinguish between variable declarations and properties with interpolation. They'd probably want, for example, to separate standard declarations from variable declarations, but not to separate standard declarations from declarations with interpolation in the property name. ??

File: src/rules/declaration-empty-line-before/index.js [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        && decl.prev().prop
+        && isStandardSyntaxProperty(decl.prev().prop)
+        && !isCustomProperty(decl.prev().prop)) {
+        expectEmptyLineBefore = !expectEmptyLineBefore
+      }
+
+      // Check for at least one empty line
+      const before = decl.raws["before"]
+      const emptyLineBefore = before.indexOf("\n\n") !== -1
+        || before.indexOf("\n\r\n") !== -1
```
Comment (davidtheclark) [inline src/rules/declaration-empty-line-before/index.js]: Should we extract out a `hasEmptyLineBefore` util? It seems this pattern is being repeated quite a bit.

File: src/rules/declaration-empty-line-before/index.js (hunk 1 above)
Comment (jeddy3) [inline src/rules/declaration-empty-line-before/index.js]: I'm in favour of pushing anything like that to a plugin. Even if it means duplicating logic within it. *Consistently* distancing, by using the `isStandardSyntax*` utils, the core rules from non-standard syntax feels more sustainable and feels like we're keeping that can of worms securely sealed.

File: src/rules/declaration-empty-line-before/index.js (hunk 2 above)
Comment (jeddy3) [inline src/rules/declaration-empty-line-before/index.js]: Good idea. Done in https://github.com/stylelint/stylelint/pull/1694/commits/d5e9ba293f26c8920ecae3c247c94c1cd09d090d

File: CHANGELOG.md [hunk 3]
```diff
@@ -2,6 +2,8 @@
 
 - Added: `block-closing-brace-empty-line-before` rule.
 - Added: `custom-property-empty-line-before` rule.
+- Added: `declaration-empty-line-before` rule.
+- Added: `media-feature-name-case` rule.
```
Comment (ntwb) [inline CHANGELOG.md]: This isn't supposed to be in this PR, was added in https://github.com/stylelint/stylelint/commit/c1dd30891aa061dc39382f70ff7b2454087116fc

Comment (jeddy3) [inline CHANGELOG.md]: My bad. Sloppy rebase.

File: src/rules/declaration-empty-line-before/index.js (hunk 1 above)
Comment (davidtheclark) [inline src/rules/declaration-empty-line-before/index.js]: I'm not sure I agree here. Maybe I'm misunderstanding, but it looks to me like we're currently adding code to deliberately ignore non-standard syntax that is not problematic. Given the infinite possibilities for non-standard syntax, I don't think we can reliably hold to a principle like "Consistently ignore non-standard syntax". I think the best we can do is a more pragmatic one, like "Ignore problematic non-standard syntax". By that principle, we wouldn't bother *adding code* to ignore non-standard syntax that we don't anticipate causing problems. And we may make a little accommodation that addresses non-standard syntax because we see that one line of code will reduce fuss and improve UX.

Comment (jeddy3) [inline src/rules/declaration-empty-line-before/index.js]: Ok, I'm convinced.

We can drop the use of `isStandardSyntaxProperty` here. And then add

```js
// SCSS var (e.g. $var: x), list (e.g. $list: (x)) or map (e.g. $map: (key:value))
if (decl.prop[0] === "$") { return false }

// Less var (e.g. @var: x)
if (decl.prop[0] === "@") { return false }
```

To `isStandardSyntaxDeclaration`. That will ignore SCSS-like variables, maps etc and Less variables (but not declaration that use variable *interpolation* in their property name).

**Comment by jeddy3** [comment]
@davidtheclark Done. Seems to work well. I had to adapt

```js
// Less var (e.g. @var: x)
if (decl.prop[0] === "@") { return false }
```

To:

```js
// Less var (e.g. @var: x), but exclude variable interpolation (e.g. @{var})
if (prop[0] === "@" && prop[1] !== "{") { return false }
```

To account for Less interpolation at the start of a string. Other than that, seems good.

**Comment by davidtheclark** [comment]
Nicely done @jeddy3!
